### DISCUSSION TITLE
Test ending before running all sequences

### DISCUSSION THREAD

**Discussion by ghost** [pyuvm/pyuvm#185] [Q&A] [answered]
So I have an agent that I named *ctrl_env* and I wrapped around an environment.  
It has a driver, a monitor, a sequencer and an analysis port.  
The driver run phase looks something like this:

```python
    self.bfm.start()
    while True:
            seq_item = await self.seq_item_port.get_next_item()
            await self.bfm.send_item(seq_item)
            self.seq_item_port.item_done()
```

In the BFM, the *send_item* function awaits to put the given item in the driver_queue. The background driver function of the BFM is as follows:

```python
async def driver(self):
        while True:
            await RisingEdge(self.dut.clk)
            seq_item = await self.driver_queue.get()
            self.dut.signal_i.value = seq_item.signal_i
```

I have a simple sequence in which, for N cycles, I change drive the value of *signal_i* to a random number. In each iteration I create a new item and  run _start_item_ and _finish_item_.

The main problem is that the sequencer is generating all the sequence items, the driver receives them and they are put in the driver_queue of the BFM but I am missing another clock cycle.  

For example, let's say I have five sequence items being generated with the following values for *signal_i*: [1, 2, 3, 4, 5]. I am expecting to finish the simulation on a RisingEdge with the value for *signal_i* being changed from 4 to 5, that is; just ready to drive the last sequence item.  
The thing is, what I am seeing is that the simulation ends on a RisingEdge ready to drive the sequence item with the 4 value, not 5; there is a clock cycle missing.

The reset is performed as shown in the TinyALU example.

Is there anything I am missing? I know for sure that it is probably something I coded the wrong way but it has been a couple of days trying to debug it and still I have found no solution. If more information on the setup is needed I can provide more. Thanks in advance!

EDIT:  
Maybe is not relevant but the monitor BFM function awaits FallingEdges and the driver function awaits Rising ones.  
Also, here is what I consider relevant time logs. As you can see, all the items send by the sequencer arrive to the Driver and they also make it to the BFM driver_queue, but the last one is not being driven.

```
1.00ns BFM........Waiting for item in driver_queue  
1.00ns DRIVER.....Got sequence item from sequencer: signal_i=1  
1.00ns BFM........Item put on driver queue: signal_i=1  
1.00ns BFM........Got sequence item from queue and driving it: signal_i=1  
1.00ns DRIVER.....Got sequence item from sequencer: signal_i=2  
1.00ns BFM........Item put on driver queue: signal_i=2  
1.00ns DRIVER.....Got sequence item from sequencer: signal_i=3  

2.00ns BFM........Waiting for item in driver_queue  
2.00ns BFM........Got sequence item from queue and driving it: signal_i=2  
2.00ns BFM........Item put on driver queue: signal_i=3  
2.00ns DRIVER.....Got sequence item from sequencer: signal_i=4

3.00ns BFM........Waiting for item in driver_queue  
3.00ns BFM........Got sequence item from queue and driving it: signal_i=3  
3.00ns BFM........Item put on driver queue: signal_i=4  
3.00ns DRIVER.....Got sequence item from sequencer: signal_i=5

4.00ns BFM.....Waiting for item in driver_queue  
4.00ns BFM.....Got sequence item from queue and driving it: data_i=4  
4.00ns BFM.....Item put on driver queue: signal_i=5
END OF SIM```

**Comment by raysalemi** [comment]
I'm guessing that you're dropping the objection in the `run_phase()` after you send the 5th item, but the testbench hasn't had time to digest that item.

**Reply by ghost** [reply]
That makes sense. Is there anything I could do to prevent that? Checking if there are items to be driven in the driver_queue? What do you think will be the best approach? Thanks!

**Reply by raysalemi** [reply, marked as answer]
That is a never-ending question. Nobody has come up with a consistent solution. Sometimes it makes sense to just wait an appropriate amount of time.

One approach that might be nice in Python is to create an awaitable object and put it in the transaction.  Then have the monitor release the await in every transaction it sees.
